### PR TITLE
Update resume feature

### PR DESCRIPTION
**Pull request by ktindiana**
Call resume with a single flag (issue https://github.com/ktindiana/sphinxval/issues/73)
Resume function does not overwrite SPHINX_dataframe file in a dangerous way (issue https://github.com/ktindiana/sphinxval/issues/9)
Resume function fully implemented and tested. Resume now checks and removes for duplicate
forecast-observation pairs. (issue https://github.com/ktindiana/sphinxval/issues/41)
There is no restriction on forecast timing.

Name of json file being read in is listed prior to the load function and
warning returned when json.load results in an empty dictionary. (issue https://github.com/ktindiana/sphinxval/issues/105)

Captured command line call to sphinx at beginning of log. (issue https://github.com/ktindiana/sphinxval/issues/102)

All matching information is now in the SPHINX_dataframe (issue https://github.com/ktindiana/sphinxval/issues/8).
The Match Reports have been removed from the logs.

**Note:** In extract_time_forecast_type, the syntax below is deprecated, per a warning provided by pandas:
sel_df.loc[len(sel_df)] = row.values
This call ended up changing the datatype of the columns in sel_df if a None of NaT value was present in row. Needed to explicitly cast datetime rows extracted from columns of sel_df to avoid a crash. This warning continues to appear, but does not impact other validated quantities.

### CODE REVIEW

File: sphinxval/sphinx/sphinx.py [hunk 1]
```diff
@@ -64,11 +64,17 @@ def validate(data_list, model_list, top=None, DoResume=False, df_pkl=""):
             
     """
     setup_logging()
+    
+    #Reconstruct command line execution of sphinx
```
Comment (rickyegeland) [inline sphinxval/sphinx/sphinx.py:68]: This should be done much more simply with sys.argv https://docs.python.org/3/library/sys.html#sys.argv

File: bin/sphinx.py [hunk 2]
```diff
@@ -11,50 +11,23 @@
         help=("If json files and time profile txt files not in the same directory, "
             "specify the base path in which the model json and time profile txt files "
             "might be found. Optional. "))
-parser.add_argument("--resume",
-        help=("Specify resume to add forecasts to an existing "
-            "dataframe and recalculate metrics."), action="store_true")
-parser.add_argument("--DataFrame", type=str, default="",
-        help=("Filename of existing Pandas DataFrame containing the results of a previous SPHINX run in pkl format. "
-            "Must specify if choose resume."))
+parser.add_argument("--resume", type=str, default="",
```
Comment (rickyegeland) [inline bin/sphinx.py:14]: None is a better default for an unset command line argument, or an unset keyword argument in general.

File: bin/sphinx.py [hunk 3]
```diff
@@ -11,50 +11,23 @@
...
-#report_only = args.ReportOnly
-
-
-#show_plot = args.showplot
-#PrintToScreen = args.PrintToScreen
-#write_report = args.WriteReport
-#SaveMetrics = args.SaveMetrics
-#onemodel = args.OneModel
+df_pkl = args.resume
+resume = False
```
Comment (rickyegeland) [inline bin/sphinx.py:28]: It doesn't look like a separate boolean is required.  Just check the existence of args.resume to determine if resume should be done.

File: bin/sphinx.py [hunk 4]
```diff
@@ -11,50 +11,23 @@
...
+if args.resume != "":
+    resume = True
 
 
 sphinxval.sphinx.validate(data_list, model_list, top, resume, df_pkl)
```
Comment (rickyegeland) [inline bin/sphinx.py:33]: I suggest defining an optional keyword argument for the resume feature, then calling it like this:

```python
phinxval.sphinx.validate(data_list, model_list, top, resume=args.resume)
```

File: sphinxval/sphinx/sphinx.py (hunk 1 above)
Comment (rickyegeland) [reply sphinxval/sphinx/sphinx.py:68]: You don't need to keep track of all your own arguments.

File: bin/sphinx.py (hunk 4 above)
Comment (rickyegeland) [reply bin/sphinx.py:33]: ... then if the argument is set with a dataframe, it is implied that resume should be enabled and use that dataframe.  This eliminates an extra variable to track.

File: requirements.txt [hunk 5]
```diff
@@ -12,3 +12,9 @@ markdown==3.5
 PyPDF2==3.0.1
 gitpython==3.1.43
 beautifulsoup4==4.12.3
+#If use PyTables
```
Comment (rickyegeland) [inline requirements.txt:15]: I don't see PyTables, tables used below.  Can't we eliminate these new requirements?

File: sphinxval/utils/resume.py [hunk 6]
```diff
@@ -129,76 +129,76 @@ def last_prediction_windows(df):
...
-                " and energy channel " + ek + " last prediction "
-                "window start " + str(ref_pred_win_st))
-            logger.debug("Current forecast prediction window "
-                    + str(pred_win_st))
-            if pred_win_st > ref_pred_win_st:
-                logger.debug("Keeping forecast")
-                select_objs[ek].append(obj)
-    
-    return select_objs
+#def check_fcast_for_resume(df, model_objs):
```
Comment (rickyegeland) [inline sphinxval/utils/resume.py:132]: Just delete this huge block of commented out code.  Git will save it if you ever want it back.

File: sphinxval/utils/validation.py [hunk 7]
```diff
@@ -426,7 +526,9 @@ def prepare_outdirs():
 def write_df(df, name, verbose=True):
     """Writes a pandas dataframe to the standard location in multiple formats
     """
-    dataformats = (('pkl',  getattr(df, 'to_pickle'), {}),
+    dataformats = (#('parquet',  getattr(df, 'to_parquet'),  dict(compression='gzip')),
```
Comment (rickyegeland) [inline sphinxval/utils/validation.py:529]: Get rid of commented out experimental data formats.

**Review by rickyegeland — CHANGES_REQUESTED** [review]
Just a couple of simple changes requested.  Removing extra requirements is more important.

File: sphinxval/sphinx/sphinx.py (hunk 1 above)
Comment (ktindiana) [reply sphinxval/sphinx/sphinx.py:68]: logger.info("SPHINX called with:",  sys.argv) did not work. It passed a logging error:
 File "/Users/kwhitman/Documents/Programs/Validation/SPHINX/sphinxval/sphinxval/sphinx/sphinx.py", line 66, in validate
    logger.info("SPHINX called with:", sys.argv)
Message: 'SPHINX called with:'
Arguments: (['bin/sphinx.py', '--ModelList', 'model/SEPVAL2023_Europe/SEPVAL2023_Europe_all_20240806_NoAMSU_2.list', '--DataList', 'lists/obs_GOES_SEPVAL2023.list', '--Resume', 'output/pkl/SPHINX_dataframe.pkl'],)

Also, the logger didn't like the comma separated inputs. I had to convert to a string. But having the sys.argv variable made it much easier!

File: bin/sphinx.py (hunk 3 above)
Comment (ktindiana) [reply bin/sphinx.py:28]: less changes than expected. Changed variable to Resume to differentiate between resume.py that gets imported as a module.

File: bin/sphinx.py (hunk 4 above)
Comment (ktindiana) [reply bin/sphinx.py:33]: apologies for being thick

**Pull request merged**